hasktorch/src/Torch/Typed/NN/Recurrent/LSTM.hs

```diff
@@ -0,0 +1,478 @@
+import           Torch.Typed.Factories
+import           Torch.Typed.Native       (expand, lstm)
+import           Torch.Typed.NN
+import           Torch.Typed.Tensor
+
+
+-- | A specification for a long, short-term memory layer.
+--
+data LSTMSpec device (dir :: Directionality)
+        (dtype :: D.DType)
+        (numLayers:: Nat)
+        (inputDim :: Nat)
+        (hiddenDim :: Nat) =
+    LSTMSpecZerosInit DropoutSpec -- ^ Weights drawn from Xavier-Uniform with zeros-value
+                                  --   initialized biases and cell states.
+    | LSTMSpec                        -- ^ Weights drawn from Xavier-Uniform
+                                      --   with zeros-value initialized biases
+                                      --   and user-provided cell states.
+        (Tensor device dtype '[numLayers * (NumberOfDirections dir), hiddenDim]) -- ^ The initial values of the memory cell
+        (Tensor device dtype '[numLayers * (NumberOfDirections dir), hiddenDim]) -- ^ The initial values of the hidden state
+        DropoutSpec
+
+    | LSTMSpecLearnedInit             -- ^ Weights drawn from Xavier-Uniform
+                                      --   with zeros-value initialized biases
+                                      --   and learned cell states.
+        (Tensor device dtype '[numLayers * (NumberOfDirections dir), hiddenDim]) -- ^ The initial (learnable)
+                                      -- values of the memory cell
+        (Tensor device dtype '[numLayers * (NumberOfDirections dir), hiddenDim]) -- ^ The initial (learnable)
+                                      -- values of the hidden state
+        DropoutSpec
+    deriving (Show, Generic)
+
+
+-- | LSTMParams
+-- Input-to-hidden, hidden-to-hidden, and bias parameters for a mulilayered
+-- (and optionally) bidirectional LSTM.
+--
+data LSTMParams (device :: (D.DeviceType, Nat)) (dtype :: D.DType) (numDirections :: Nat) (inputDim :: Nat) (hiddenSize :: Nat) (numLayers :: Nat) where
+    LSTMLayer1 ::Parameter device dtype '[4 * hiddenSize, inputDim]
+                    -> (Parameter device dtype '[4 * hiddenSize, hiddenSize])
+                    -> (Parameter device dtype '[4 * hiddenSize ])
+                    -> (Parameter device dtype '[4 * hiddenSize ])
+                    -> LSTMParams device dtype numDirections inputDim hiddenSize 1
+    LSTMLayerK ::(1 <= numLayers)
+                    => LSTMParams device dtype numDirections inputDim hiddenSize numLayers
+                    -> (Parameter device dtype '[4 * hiddenSize, numDirections * hiddenSize])
+                    -> (Parameter device dtype '[4 * hiddenSize, hiddenSize])
+                    -> (Parameter device dtype '[4 * hiddenSize ])
+                    -> (Parameter device dtype '[4 * hiddenSize ])
+                    -> LSTMParams device dtype numDirections inputDim hiddenSize (numLayers + 1)
+
+--  TODO: Generics? see https://gist.github.com/RyanGlScott/71d9f933e823b4a03f99de54d4b94d51
```

Comment:
I don't think you'll be able to derive a generics instance for `LSTMParams` because of the constraint in the data constructor for `LSTMLayerK`

Reply:
oh, https://gist.github.com/RyanGlScott/71d9f933e823b4a03f99de54d4b94d51 is neat

Reply:
Yes, but a finicky, error-prone investment at this point.  Esp if you want to move everything to `HList`s later.